mongodb/src/main/java/org/hibernate/ogm/datastore/mongodb/MongoDBDialect.java

```diff
@@ -1733,6 +1742,114 @@ private static ReadPreference getReadPreference(AssociationContext associationCo
+		StringBuilder commandLine = createCallStoreProcedureCommand( storedProcedureName, params );
+		Document result = callStoredProcedure( commandLine );
+		Object resultValue = result.get( "retval" );
+		List<Tuple> resultTuples = extractTuples( storedProcedureName, resultValue );
+		return CollectionHelper.newClosableIterator( resultTuples );
+	}
+
+	private void validate(ProcedureQueryParameters params) {
+		if ( !params.getNamedParameters().isEmpty() ) {
+			throw log.dialectDoesNotSupportNamedParametersForStoredProcedures( getClass() );
+		}
+	}
+
+	private List<Tuple> extractTuples(String storedProcedureName, Object retvalObj) {
+		if ( retvalObj instanceof Document ) {
+			Document retval = (Document) retvalObj;
+			if ( retval.size() > 1 ) {
+				throw log.multipleDocumentReturnedByStoredProcedure( storedProcedureName, retval.size() );
+			}
+			String firstRetValTag = retval.keySet().iterator().next();
+			Object firstRetVal = retval.get( firstRetValTag );
+			@SuppressWarnings("unchecked")
+			Iterable<Document> documents = (Iterable<Document>) firstRetVal;
+			List<Tuple> resultTuples = new ArrayList<>();
+			for ( Document doc : documents ) {
+				resultTuples.add( convert( doc ) );
+			}
+			return resultTuples;
+		}
+		else {
+			Tuple tuple = new Tuple();
+			tuple.put( "result", retvalObj );
+			return Collections.singletonList( tuple );
+		}
+	}
+
+	private Document callStoredProcedure(StringBuilder commandLine) {
+		try {
+			Document result = provider.getDatabase().runCommand( new Document( "$eval", commandLine.toString() ) );
+			return result;
+		}
+		catch (MongoCommandException mce) {
+			BsonDocument response = mce.getResponse();
+			throw log.unableToExecuteCommand( commandLine.toString(), response.getString( "errmsg" ).getValue(),
+					response.getString( "codeName" ).getValue(), mce );
+		}
+	}
+
+	private StringBuilder createCallStoreProcedureCommand(String storedProcedureName, ProcedureQueryParameters params) {
+		StringBuilder commandLine = new StringBuilder( storedProcedureName ).append( "(" );
+		List<Object> positionalParameters = params.getPositionalParameters();
+		for ( Object paramValue : positionalParameters ) {
+			appendStoredProcedureParamValue( storedProcedureName, commandLine, paramValue );
+			commandLine.append( "," );
+		}
+		commandLine.setLength( commandLine.length() - 1 );
+		commandLine.append( ")" );
+		return commandLine;
+	}
+
+	private void appendStoredProcedureParamValue(String storedProcedureName, StringBuilder commandLine, Object paramValue) {
+		if ( paramValue == null ) {
+			// I don't think this can happen but, just in case, null should be a valid JSON value
+			commandLine.append( "null" );
+		}
+		else {
+			if ( requiresQuotes( paramValue ) ) {
+				String escapedValue = StringHelper.escapeDoubleQuotesForJson( paramValue.toString() );
+				// I think we only have to worry about double quotes for String
+				commandLine.append( '"' ).append( escapedValue ).append( '"' );
```

Comment:
I think you also have to escape the `\` character. Otherwise a `\"` string will be changed to `"`.

Reply:
I'm ont sure I follow, but I added a test and it seems to work
Feel free to add another one before making changes, please